### ISSUE TITLE
How to bring the model into practice and deal with new data?

### ISSUE THREAD

**Issue by MaudDie** [MaartenGr/BERTopic#354]
Hi Maarten, 

First of all, kudos to the packages you created. I am impressed. 

I used your library to analyze ~ 9K free text field. I made use of 1 year of data. Now I would like to rerun this model every two weeks with (some) new data so that I gain insight into how the topics change over time, which topics get bigger/smaller which ones disappear and which ones appear. 

But the runs differ a lot. I would expect a small difference since the stochastic nature of UMAP, but I would like to see similar clusters with a slightly different dataset. A seed will not work, since the dataset will be slightly different every two weeks. I tried to work with different embedding techniques, tried the guided topic modelling technique with some given topics, and tried various parameters of UMAP and HDBSCAN. Your guided topic modelling works great of the model is fitted, but what to do if you rerun every two weeks with a window of one year? 

Do you have any best practices to bring your model into practice? So actually see the progress of the topics with a moving window of data over time.

Thank you in advance

**Comment by MaartenGr** [comment]
A difficult problem indeed. With the stochastic nature of UMAP, it is difficult to reproduce the results, more so if you want to retrain the model with new data! There are several ways you can approach this. 

First, as you mentioned before, you can try guided topic modeling to make sure that the topics remain relatively similar between runs. Let's say you start with your 1 year of data. The basis model provides you with topic representations that can then be used as the input for guided topic modeling through `seed_topic_list`. It seems that you already started with this. Let's say you want to rerun the model two weeks later. You can use the `seed_topic_list` to make sure that at least some of the topics are found in the initial model are also found in the new model. Although new topics may then still be generated, which we typically prefer, it might inhibit the generation of new topics that might pop up two weeks later. 

Second, you train BERTopic on your 1 year of data and use the resulting `topics` as the `y` parameter in `fit_transform(y=topics)` when you retrain the model two weeks later. This forces the documents in your 1 year of data to cluster together. The great thing about this method is that any new documents can simply be assigned the label `-1` which simply refers to "I do not know how to cluster this". In other words, it is a semi-supervised approach in reproducing the results but still allowing some of the original topics to be formed. 

Third, you can retrain BERTopic without any of the steps above and use `find_topics()` to track the topic that you are interested in. Use the topic representations from the base model to search for the topics that are relevant. Note that you can also use this step combined with one of the two above as it will help you to easily track topics. 

Fourth, you can use a `random_state` with UMAP to ensure that embeddings are reduced similarly. That way, you further minimize the differences that you might have between two subsequent weeks. 

Fifth, if you believe that all interesting topics can be found in your 1 year of data, then there is no need to retrain the model. You can use `transform()` to predict which topics belong to the new document and then visualize how these have evolved over time by leveraging `topics_over_time` and `visualize_topics_over_time`. 

This turned out a bit longer than intended but hopefully, this helps you out! If not, please let me know.

**Comment by MaartenGr** [comment]
Due to inactivity, I am going to close this for now. If you want to continue this discussion, please let me know and I'll make sure to reopen the issue.
